source/mqtt3_to_mqtt5_adapter.c

```diff
@@ -22,13 +51,61 @@ struct aws_mqtt_client_connection_5_impl {
     struct aws_mqtt5_listener *listener;
     struct aws_event_loop *loop;
 
-    struct aws_mutex lock;
+    /*
+     * An event-loop-internal flag that we can read to check to see if we're in the scope of a callback
+     * that has already locked the adapter's lock.  Can only be referenced from the event loop thread.
+     *
+     * We use the flag to avoid deadlock in a few cases where we can re-enter the adapter logic from within a callback.
+     * It also provides a nice solution for the fact that we cannot safely upgrade a read lock to a write lock.
+     */
+    bool in_synchronous_callback;
+
+    /*
+     * What state the mqtt311 interface wants to be in.  Always either AWS_MCS_CONNECTED or AWS_MCS_STOPPED.
+     * Tracked separately from the underlying mqtt5 client state in order to provide a coherent view that is
+     * narrowed solely to the adapter interface.
+     */
+    enum aws_mqtt_adapter_state adapter_state;
```

Comment:
should this enum be `aws_mqtt5_client_state`?

Reply:
nvm, this is used below intentionally as `aws_mqtt_adapter_state`. Maybe the description needs to be changed to reflect the enum values available for it( `AWS_MQTT_AS_STAY_CONNECTED` instead of `AWS_MCS_CONNECTED` )? Or clarified that the mqtt5 client is always in either the listed `AWS_MCS_CONNECTED` or `AWS_MCS_STOPPED` desired states? Trivial but just for clarification.

Reply:
Updated the out-of-date comment